### PR TITLE
Raise a NameError if a field name shadows a BaseModel attribute

### PR DESCRIPTION
**Pull request by nabla-c0d3**
This is a PR for #242. I was not sure about the right place to add this check (and its test) so let me know if there's a better way.

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -112,6 +112,12 @@ def __new__(mcs, name, bases, namespace):
         annotations = namespace.get('__annotations__', {})
         # annotation only fields need to come first in fields
         for ann_name, ann_type in annotations.items():
+            # Ensure that the field's name does not shadow an existing attribute
+            for base in reversed(bases):
```
Comment (samuelcolvin) [inline pydantic/main.py]: You'll need to check for non-annotation attributes too, eg. line 131 below. Best to make this a method so the logic can be reused.

File: tests/test_schema.py [hunk 2]
```diff
@@ -406,3 +406,9 @@ class Model(BaseModel):
             },
         },
     }
+
+
+def test_field_name_shadows_attribute():
```
Comment (samuelcolvin) [inline tests/test_schema.py]: this is the wrong place, I guess move to `test_main.py`. You should also add a test for a non annotation only attribute.

File: pydantic/main.py (hunk 1 above)
Comment (nabla-c0d3) [reply pydantic/main.py]: What are non-annotation attributes? Is it the ones created with the create_model() function?

Comment (samuelcolvin) [reply pydantic/main.py]: No they're attributes with a value, eg. `foo = 123`. Read the code on the line I suggested.

**Comment by nabla-c0d3** [comment]
All done?

File: pydantic/main.py [hunk 3]
```diff
@@ -141,6 +143,16 @@ def __new__(mcs, name, bases, namespace):
         }
         return super().__new__(mcs, name, bases, new_namespace)
 
+    @staticmethod
+    def _validate_field_name(bases: List[Type['BaseModel']], field_name: str) -> None:
+        """
+        Ensure that the field's name does not shadow an existing attribute of the model.
+        """
+        for base in reversed(bases):
+            if getattr(base, field_name, None):
```
Comment (samuelcolvin) [inline pydantic/main.py]: Is reversed required?

File: pydantic/main.py [hunk 4]
```diff
@@ -141,6 +143,16 @@ def __new__(mcs, name, bases, namespace):
         }
         return super().__new__(mcs, name, bases, new_namespace)
 
+    @staticmethod
+    def _validate_field_name(bases: List[Type['BaseModel']], field_name: str) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py]: I don't think this needs to be a static method on the class. Just a simple method in utils would be better.

**Comment by nabla-c0d3** [comment]
Ok it should have everything now but let me know.

**Comment by samuelcolvin** [comment]
awesome, thank you very much.

**Pull request merged**